core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/SearchResultGenerator.kt

```diff
@@ -32,23 +32,30 @@ class ZimSearchResultGenerator @Inject constructor(
   private val sharedPreferenceUtil: SharedPreferenceUtil,
   private val zimReaderContainer: ZimReaderContainer
 ) : SearchResultGenerator {
+  private var searchTerm = ""
+
   override fun generateSearchResults(searchTerm: String) =
-    if (searchTerm.isNotEmpty()) readResultsFromZim(searchTerm)
+    if (searchTerm.also { this.searchTerm = it }.isNotEmpty()) readResultsFromZim()
     else emptyList()
 
-  private fun readResultsFromZim(it: String) =
+  private fun readResultsFromZim() =
     if (sharedPreferenceUtil.prefFullTextSearch)
-      zimReaderContainer.search(it, 200).run { fullTextResults() }
+      zimReaderContainer.search(searchTerm, 200).run { fullTextResults() }
     else
-      zimReaderContainer.searchSuggestions(it, 200).run { suggestionResults() }
+      zimReaderContainer.searchSuggestions(searchTerm, 200).run { suggestionResults() }
 
   private fun fullTextResults() = generateSequence {
     zimReaderContainer.getNextResult()?.title?.let(::ZimSearchResultListItem)
   }.filter { it.value.isNotBlank() }
     .toList()
 
-  private fun suggestionResults() = generateSequence {
-    zimReaderContainer.getNextSuggestion()?.let { ZimSearchResultListItem(it.title) }
-  }.distinct()
-    .toList()
+  private fun suggestionResults() =
+    generateSequence {
+      zimReaderContainer.getNextSuggestion()?.let { ZimSearchResultListItem(it.title) }
+    }.distinct()
+      // Sorted results with searchTerm at top
+      .filter { it.value != searchTerm }
```

Comment:
I haven't ran the code yet but I imagine this will only emit a list with the exact search term

Reply:
This line does the exact opposite. It removes the exact search term from the list, which I add back later at the top

Reply:
sorry morning brain. Absolutely right.

I think we might be better handling this reordering in State.Results in an init block as we can play with the data a bit easier, some thing as easy as
```
(exactMatchListItem + (results - exactMatchListItem)).sortedBy{ blah }
``

Reply:
or a companion function and it can be an parameter with default argument of the function

Reply:
`State.Results` has the `values` as `List` instead of `MutableList`. If the logic was moved there, it would need to copy the list

Reply:
which I am fine with, creating a new list is not expensive for the amount of items that will be here, each map/filter creates a new list like. Immutable data should (nearly) always be preferred.

we could make the current results private or it is probably better to introduce a secondary constructor in this case because we don't need to store the intermediary representation of the results

Reply:
I ended up with `values.filter { it.value == searchString } + values.filter { it.value != searchString }`. Here, `exactMatchListItem` can't be used because it is of unknown type as the class is generic. For the `sortedBy`, I still don't know whether we want the sorting in Kotlin or in libkiwix.

To work around Kotlin data classes forcing `var`/`val` inside their constructors, I made the [companion object pretend to be one](https://stackoverflow.com/q/49561735).

Reply:
My pseudocode was a bit sloppy, this is valid

```
listOf(subItem) + superTypeList
```

I don't think we want to do any more client side sorting besides this reorder to the top. For now anyway

Reply:
This part:

> subItem

doesn't work. I can't instantiate `subItem` because we don't know the concrete class type. I am also avoiding getting the item out of the `filter` to avoid NPE/AIOOB.